v0/customerio/transform.js

```diff
@@ -0,0 +1,222 @@
+  USER_EVENT_ENDPOINT,
+  ANON_EVENT_ENDPOINT,
+  DEVICE_REGISTER_ENDPOINT,
+  DEVICE_DELETE_ENDPOINT
+} = require("./config");
+
+const deviceRelatedEventNames = [
+  "Application Installed",
+  "Application Opened",
+  "Application Uninstalled"
+];
+const deviceDeleteRelatedEventName = "Application Uninstalled";
+
+// Get the spec'd traits, for now only address needs treatment as 2 layers.
+// populate the list of spec'd traits in constants.js
+const populateSpecedTraits = (payload, message) => {
+  //console.log(message);
+  SpecedTraits.forEach(trait => {
+    const mapping = TraitsMapping[trait];
+    const keys = Object.keys(mapping);
+    keys.forEach(key => {
+      set(payload, key, get(message, mapping[key]));
+    });
+  });
+};
+
+function responseBuilder(message, evType, evName, destination) {
+  const rawPayload = {};
+  let endpoint;
+  let requestConfig = defaultPostRequestConfig;
+  //console.log(message);
+  let userId =
+    message.userId && message.userId != "" ? message.userId : undefined;
+
+  const response = defaultRequestConfig();
+  response.userId = message.userId ? message.userId : message.anonymousId;
+
+  if (evType === EventType.IDENTIFY) {
+    // populate speced traits
+    populateSpecedTraits(rawPayload, message);
+    if (message.context.traits) {
+      const traits = Object.keys(message.context.traits);
+      traits.forEach(trait => {
+        //populate keys other than speced traits
+        if (!SpecedTraits.includes(trait)) {
+          set(rawPayload, trait, get(message, "context.traits." + trait));
+        }
+      });
+    }
+
+    if (message.user_properties) {
+      const userProps = Object.keys(message.user_properties);
+      userProps.forEach(prop => {
+        let val = get(message, "user_properties." + prop);
+        // send only top level keys
+        if (isPrimitive(val)) {
+          set(rawPayload, prop, val);
+        }
+      });
+    }
+
+    set(
+      rawPayload,
+      "created_at",
+      Math.floor(new Date(message.originalTimestamp).getTime() / 1000)
+    );
+    //console.log(rawPayload);
+
+    if (userId) {
+      endpoint = IDENTITY_ENDPOINT.replace(":id", userId);
+    } else {
+      return { statusCode: 400, error: "userId not present" };
+    }
+    requestConfig = defaultPutRequestConfig;
+  } else {
+    if (message.properties) {
+      // use this if only top level keys are to be sent
+
+      /* const eventProps = Object.keys(message.properties);
+      eventProps.forEach(prop => {
+        let val = get(message, "properties." + prop);
+        if (isPrimitive(val)) {
+          set(rawPayload.data, prop, val);
+        }
+      }); */
+
+      if (deviceDeleteRelatedEventName == evName) {
+        const token = get(message, "context.device.token");
+        if (userId && token) {
+          endpoint = DEVICE_DELETE_ENDPOINT.replace(":id", userId).replace(
+            ":device_id",
+            token
+          );
+
+          response.endpoint = endpoint;
```

Comment:
Setting everything together is more readable.

{
  endpoint,
  method: "DELETE",
  headers: {
  }
}

Reply:
Left it as it is now. Once we come back here for someone, will do a round of code cleanup.